RFCs/0000-pdk-template-v2.md

```diff
@@ -0,0 +1,296 @@
+| evaluate_ruby_files | An array of files to [evaluate](https://ruby-doc.org/core-2.6.4/Kernel.html#method-i-eval) prior to rendering ERB templates. This is used to inject complex logic into the templating system without the troubles of having complex logic in ERB.  For example complex branching based on settings. Default is empty array |
+| default_settings | A YAML hash of the default settings for the template. Default is empty hash |
+| always_apply | Always apply this template. If not specified in the template list, it will be applied last. Default is `false` |
+| tags | An array of string tags for this template. Always includes the directory name of the module. Default is an empty array |
+
+### `<name>/template_schema.json`
+
+### Example templates
+
+#### A template for a default module file
+
+The `.pdkignore` file will be used in all PDK compatible modules therefore it should always be applied.
+
+File layout
+
+``` text
+pdkignore/
+  +- template.yml
+  +- template_schema.json
+  +- files/
+       +- .pdkignore.erb
+```
+
+
+`template.yml`
+
+``` yaml
+---
+name: "PDK Ignore"
+description: >-
+  The PDK Ignore template manages the .pdkignore file for Puppet Modules
+type: module
+always_apply: true
+tags: ignore default
+default_settings:
+  required:
+    - '/appveyor.yml'
+    - '/.fixtures.yml'
+    - '/Gemfile'
+    - '/.gitattributes'
+    - '/.gitignore'
+    - '/.gitlab-ci.yml'
+    - '/.pdkignore'
+    - '/Rakefile'
+    - '/rakelib/'
+    - '/.rspec'
+    - '/.rubocop.yml'
+    - '/.travis.yml'
+    - '/.yardopts'
+    - '/spec/'
+    - '/.vscode/'
+    # ... many other paths
+  paths: ~
+```
+
+`template_schema.json`
+
+``` json
+{
+  "$schema": "http://json-schema.org/draft-06/schema#",
+  "$id": "http://puppet.com/schema/does_not_exist.json",
+  "type": "object",
+  "title": "The PDK Ignore Template Schema",
+  "properties": {
+    "paths": {
+      "$id": "#/properties/paths",
+      "title": "Paths property",
+      "description": "Array of file paths to also ignore by the PDK, for example, when building a module",
+      "type": "array",
+      "items": {
+        "type": "string"
+      },
+    },
+  },
+  "definitions": {}
+}
+```
+
+**Note** that the `required` setting is not documented, or has a schema, as it's internal to the template.
+
+`.pdkignore.erb`
+
+Same as [V1 template](https://github.com/puppetlabs/pdk-templates/blob/90e5efab878facd18f838755d9e95b226b1e0776/moduleroot/.pdkignore.erb).
+
+#### An opt-in template for CI configuration
+
+> TODO
+
+#### A template only for new object
+
+
+### Changes to `.sync.yml`
+
+Each module may contain a `.sync.yml` file which contains the module specific settings when rendering the templates. For the new templates, the .sync.yml will have additional settings, which will also be backwards compatible with the V1 templating engine.
```

Comment:
How does this align with https://github.com/puppetlabs/pdk-planning/blob/master/RFCs/0002-add-pdk-config.md ?

Reply:
Yeah, I would still like to explore more user-friendly and discoverable means of customizing the templates other than `.sync.yml`, and it seems like a Template System v2 would be a good opportunity to do that.

Reply:
I would separate these two concerns.

`.sync.yml` is just a storage mechanism for module level customisations. (Not that I would), but this could also be an XML file, or compressed binary file, or even a Google Doc.

Editing the content of that file is the realm of the `pdk config` command. The UX for that is not coupled to the storage location

Granted that YAML is not the best language, but it's certainly not the worst and there are many libraries in many languages to help make editing this a better experience.  Combined with the JSON Schema validation it would be possible for tools to _consume_ the template metadata, and create a really compelling experience.

Reply:
I'm not opposed to it being YAML, I just specifically think that the (hidden on POSIX platforms) filename `.sync.yml` is a bad way to convey what that file actually does these days. The naming is an artifact of the file's evolution from `modulesync`.

Even if it looked very similar inside but was called `module_config.yml` or something I think that would be a big improvement.

Reply:
As far as whether or not that change belongs as part of this RFC, I can go either way. I was just thinking that if the format of the file needed to change at all anyway, it might be easier to do the rename at the same time.